board/safety/safety_mazda.h

```diff
@@ -21,41 +22,86 @@
 #define MAZDA_DRIVER_TORQUE_FACTOR 1
+#define MAZDA_MAX_TORQUE_ERROR 350
 
-int mazda_cruise_engaged_last = 0;
-int mazda_rt_torque_last = 0;
-int mazda_desired_torque_last = 0;
-uint32_t mazda_ts_last = 0;
-struct sample_t mazda_torque_driver;         // last few driver torques measured
+const CanMsg MAZDA_TX_MSGS[] = {{MAZDA_LKAS, 0, 8}};
+int mazda_lkas_on = 0;
 
 // track msgs coming from OP so that we know what CAM msgs to drop and what to forward
 static int mazda_rx_hook(CAN_FIFOMailBox_TypeDef *to_push) {
   int bus = GET_BUS(to_push);
   int addr = GET_ADDR(to_push);
 
-  if ((addr == MAZDA_STEER_TORQUE) && (bus == MAZDA_MAIN)) {
-    int torque_driver_new = GET_BYTE(to_push, 0) - 127;
-    // update array of samples
-    update_sample(&mazda_torque_driver, torque_driver_new);
-  }
+  if (bus == MAZDA_MAIN) {
 
-  // enter controls on rising edge of ACC, exit controls on ACC off
-  if ((addr == MAZDA_CRZ_CTRL) && (bus == MAZDA_MAIN)) {
-    int cruise_engaged = GET_BYTE(to_push, 0) & 8;
-    if (cruise_engaged != 0) {
-      if (!mazda_cruise_engaged_last) {
-        controls_allowed = 1;
+    // sample speed
+    if (addr == MAZDA_WHEEL_SPEED) {
+      int speed = 0;
+      // sum 4 wheel speeds
+      for (int i=0; i<8; i+=2) {
+        int next_byte = i + 1;  // hack to deal with misra 10.8
+        speed += (GET_BYTE(to_push, i) << 8) + GET_BYTE(to_push, next_byte) - 10000;
+      }
+      speed = speed / 4;
+      vehicle_moving = ABS(speed) > 10; // 0.1 kph
+
+      if (speed > 4950) {      // 49.5 kph
+        mazda_lkas_on = 1;
+      } else if (speed < 4450) { // 44.5 kph
+        mazda_lkas_on = 0;
+      } else {
+        // Misra-able appeasment block!
       }
     }
-    else {
-      controls_allowed = 0;
+
+    if (addr == MAZDA_STEER_TORQUE) {
+      int torque_driver_new = GET_BYTE(to_push, 0) - 127;
+      // update array of samples
+      update_sample(&torque_driver, torque_driver_new);
     }
-    mazda_cruise_engaged_last = cruise_engaged;
-  }
 
-  // if we see wheel speed msgs on MAZDA_CAM bus then relay is closed
-  if ((safety_mode_cnt > RELAY_TRNS_TIMEOUT) && (bus == MAZDA_CAM) && (addr == MAZDA_WHEEL_SPEED)) {
-    relay_malfunction_set();
+    // enter controls on rising edge of ACC, exit controls on ACC off
+    if (addr == MAZDA_CRZ_CTRL) {
+      int cruise_engaged = GET_BYTE(to_push, 0) & 8;
+      if (cruise_engaged != 0) {
+        if (!cruise_engaged_prev) {
+          // do not engage until we hit the speed at which lkas is on
+          if (mazda_lkas_on == 1) {
+            controls_allowed = 1;
+          }
+          else {
+            controls_allowed = 0;
+            cruise_engaged = 0;
+          }
+        }
+      } else {
+        controls_allowed = 0;
+      }
+      cruise_engaged_prev = cruise_engaged;
+    }
+
+    // Exit controls on rising edge of gas press
+    if (addr == MAZDA_ENGINE_DATA) {
+      bool gas_pressed = (GET_BYTE(to_push, 4) || (GET_BYTE(to_push, 5) & 0xF0));
```

Comment:
this technically works since we're only checking if it's pressed, but let's do the shift so it's clear that it's a 12 bit signal